### PR TITLE
TIP-731- Implement webpack and remove oro requirejs bundle

### PR DESCRIPTION
**Pull request by tamarasaurus**
## Summary 

This PR removes the Oro requirejs bundle and implements webpack to compile the frontend assets (excluding less/css). 

There are a few new commands added in this PR - 
- `app/console pim:installer:dump-require-paths` - exports a file that contains the file paths for every requirejs.yml file for each bundle loaded in the app kernel (is included in the pim install command, so no need to run manually)
- `npm run webpack` - runs webpack and compiles the assets for production
- `npm run webpack-watch` - runs webpack with automatic page reload
- `npm run lint` -  runs eslint and reports errors
- `npm run lint-fix` - runs eslint and automatically fixes and reports errors
- `npm run webpack-jasmine` - compiles the js specs and runs them with karma-webpack

The only command that needs to be run to compile the frontend is `npm run webpack`

## Changes
- Update Jenkinsfile to run eslint and jasmine commands
- Add command to dump requirejs.yml paths for each bundle 
- Remove Oro requirejs bundle
- Remove Gruntfile.js
- Implement webpack
    - Add a polyfill to run dynamic requirejs calls and handle `require()` inside twig templates 
    - Add a [webpack loader to inject module config](https://github.com/akeneo/pim-community-dev/pull/6160/files#diff-14ba459a666c9096affa01dedc438346)  - instead of using `module.config()` which is not available in webpack - modules can use `__moduleConfig` to get the same data
    - Add a util module to translate the requirejs.yml config to webpack
    - Add [karma-webpack](https://github.com/webpack-contrib/karma-webpack) to run jasmine tests instead of grunt jasmine
    - Add a text-loader inside webpack config to avoid declaring `text!` when requiring html files
    - Fix all the cases where module variables are being used without being declared in script blocks in twig templates e.g. global jquery, underscore, backbone  
    - Add babel-loader with es2017 preset

## To-do
- [x] Add uglyfy/minification option to webpack.config.js
- [x] Add hot module reloading in webpack.dev.config.js 
- [x] Add command in package.json to run webpack dev server
- [x] Write migration docs listing changes
- [x] Add sourcemap 
- [x] Discuss next steps for phasing out requirejs.yml files
- [x] Move unmodified frontend libs to npm - 
    - ~src/Pim/Bundle/UIBundle/Resources/public/lib/base64/base64.js~
    - ~src/Pim/Bundle/UIBundle/Resources/public/lib/bootstrap-datetimepicker/js/bootstrap-datetimepicker.js~
    - ~src/Pim/Bundle/UIBundle/Resources/public/lib/bootstrap-switch/bootstrap.switch.js~
    - ~src/Pim/Bundle/UIBundle/Resources/public/lib/jquery/jquery-1.10.2.js~
    - ~src/Pim/Bundle/UIBundle/Resources/public/lib/jquery-numeric/jquery.numeric.js~ (not on npm)
    - ~src/Pim/Bundle/UIBundle/Resources/public/lib/json2/json2.js~
    - ~src/Pim/Bundle/UIBundle/Resources/public/lib/jstree/jquery.jstree.js~
    - ~src/Pim/Bundle/UIBundle/Resources/public/lib/jstree/jquery.hotkeys.js~ (no version on npm)
    - ~src/Pim/Bundle/UIBundle/Resources/public/lib/select2/select2.js~ (3.4.1 not available on npm)
    - ~src/Pim/Bundle/UIBundle/Resources/public/lib/underscore/underscore.js~

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | -
| Added Behats                      | OK
| Added integration tests           | -
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

**Comment by nidup** [comment]
cool! keep in mind to add a bit of info in the following file, at least, indicate the bundle we removed and maybe change the command to run to dump assets (if it has changed) https://github.com/akeneo/pim-community-dev/blob/master/UPGRADE-1.8.md#migrate-your-standard-project (i just added the upgrade file on master so you may have to rebase to get it).

**Comment by pierallard** [comment]
We have to make a documentation for the developers (core team and others) to install the needed softs like npm! Like @jjanvier does for ss.

**Comment by tamarasaurus** [comment]
@pierallard I plan to write a migration guide which will include installation instructions, changes, commands to run

File: frontend/jasmine-runner.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+var requireAll = function(requireContext) {
+    requireContext.keys().map(requireContext);
+};
+
+requireAll(require.context('./dynamic/', true, /^.*\/Bundle\/.*\/Resources\/public\/spec\/.*\/.*\.js$/));
+
+Backbone.$ = $
```
Comment (tamarasaurus) [inline frontend/jasmine-runner.js]: not necessary anymore

File: Jenkinsfile [hunk 2]
```diff
@@ -165,11 +185,11 @@ def runGruntTest() {
     node('docker') {
         deleteDir()
         try {
-            docker.image('digitallyseamless/nodejs-bower-grunt').inside("") {
+            docker.image('node').inside("") {
                 unstash "pim_community_dev_full"
-                sh "npm install"
-
-                sh "grunt"
+                sh "npm install --verbose"
```
Comment (tamarasaurus) [inline Jenkinsfile]: remove verbose

File: frontend/add-context-plugin.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
+/* eslint-env es6 */
```
Comment (tamarasaurus) [inline frontend/add-context-plugin.js]: Make folder more structured 'webpack'

File: frontend/jasmine-runner.js [hunk 4]
```diff
@@ -0,0 +1,7 @@
+var requireAll = function(requireContext) {
+    requireContext.keys().map(requireContext);
+};
+
+requireAll(require.context('./dynamic/', true, /^.*\/Bundle\/.*\/Resources\/public\/spec\/.*\/.*\.js$/));
```
Comment (tamarasaurus) [inline frontend/jasmine-runner.js]: double check regex

File: frontend/path-overrides.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+module.exports = {
```
Comment (tamarasaurus) [inline frontend/path-overrides.js]: put in enrich bundle

File: frontend/requirejs-utils.js [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+                const contents = fs.readFileSync(yaml, 'utf8')
+                const parsed = getYaml.parse(contents)
+                const requirePaths = parsed.config.paths || {}
+                const requireMaps = _.get(parsed.config, 'map.*') || {}
+                const mergedPaths = Object.assign(requirePaths, requireMaps)
+                const absolutePaths = utils.getAbsolutePaths(mergedPaths, yaml)
+
+                paths = deepMerge(paths, absolutePaths)
+                config = deepMerge(config, parsed.config.config || {})
+            } catch (e) {}
```
Comment (tamarasaurus) [inline frontend/requirejs-utils.js]: think about logging here

File: frontend/requirejs-utils.js [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+    /**
+     * Gets the absolute path of a module and returns an object containing a map of module names to paths
+     * Example: pimui/js/pim-formupdatelistener -> {user home}/src/Pim/Bundle/UIBundle/Resources/public/js/pim-formupdatelistener'
+     * @param  {Object} paths    An object containing module/path mapping
+     * @param  {String} yamlFilename The filename of the requirejs.yaml file
+     * @return {Object}          An object containing mapping of module name to absolute file location
+     */
+    getAbsolutePaths(paths, yamlFilename) {
+        const absolutePaths = {}
```
Comment (tamarasaurus) [inline frontend/requirejs-utils.js]: load files from web/bundles instead

File: package.json [hunk 8]
```diff
@@ -22,5 +51,11 @@
   "license": "OSL-3.0",
   "bugs": {
     "url": "https://github.com/akeneo/pim-community-dev/issues"
+  },
+  "dependencies": {
+    "bundle-loader": "^0.5.5",
+    "karma-es6-shim": "^1.0.0",
+    "merge-objects": "^1.0.5",
+    "pascal-case": "^2.0.1"
```
Comment (tamarasaurus) [inline package.json]: remove pascal case

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/pim-init.js [hunk 9]
```diff
@@ -1,10 +1,45 @@
...
+        WidgetContainer,
+        LastOperationsWidget,
+        CompletenessWidget,
+        PatchFetcher,
+        RequirePolyfill
+    ) {
         'use strict';
         var initialized = false;
 
+        // console.log('Add RequirePolyfill', RequirePolyfill)
```
Comment (tamarasaurus) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/pim-init.js]: remove

File: src/Pim/Bundle/InstallerBundle/Command/DumpRequirePathsCommand.php [hunk 10]
```diff
@@ -0,0 +1,72 @@
+<?php
+
+/*
+ * This file is part of the Akeneo PIM Enterprise Edition.
+ *
+ * (c) 2017 Akeneo SAS (http://www.akeneo.com)
```
Comment (tamarasaurus) [inline src/Pim/Bundle/InstallerBundle/Command/DumpRequirePathsCommand.php]: fix the license - remove this part and put the correct license

File: src/Pim/Bundle/InstallerBundle/Command/DumpRequirePathsCommand.php [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * Dump a file called require-paths containing all requirejs.yml config files for reach registered bundle
+ *
+ * @author Tamara Robichet <tamara.robichet@akeneo.com>
+ */
+class DumpRequirePathsCommand extends ContainerAwareCommand
+{
+
```
Comment (tamarasaurus) [inline src/Pim/Bundle/InstallerBundle/Command/DumpRequirePathsCommand.php]: :fire:

File: src/Pim/Bundle/InstallerBundle/Command/DumpRequirePathsCommand.php [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $output->writeln('Generating require.js main config');
+
+        $webRoot = 'web';
```
Comment (tamarasaurus) [inline src/Pim/Bundle/InstallerBundle/Command/DumpRequirePathsCommand.php]: get from configuration

File: src/Pim/Bundle/InstallerBundle/Command/DumpRequirePathsCommand.php [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+        $mainConfigContent = json_encode($this->collectConfigPaths(), JSON_UNESCAPED_SLASHES);
+
+        $mainConfigContent = 'module.exports = ' . $mainConfigContent;
+        $mainConfigContent = str_replace(',', ",\n", $mainConfigContent);
+        $mainConfigFilePath = $webRoot . DIRECTORY_SEPARATOR . self::MAIN_CONFIG_FILE_NAME;
+        if (false === @file_put_contents($mainConfigFilePath, $mainConfigContent)) {
```
Comment (tamarasaurus) [inline src/Pim/Bundle/InstallerBundle/Command/DumpRequirePathsCommand.php]: remove @

File: src/Pim/Bundle/InstallerBundle/Command/DumpRequirePathsCommand.php [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+            throw new \RuntimeException('Unable to write file ' . $mainConfigFilePath);
+        }
+    }
+
+    public function collectConfigPaths()
```
Comment (tamarasaurus) [inline src/Pim/Bundle/InstallerBundle/Command/DumpRequirePathsCommand.php]: add php doc

File: src/Pim/Bundle/InstallerBundle/Command/DumpRequirePathsCommand.php (hunk 12 above)
Comment (tamarasaurus) [inline src/Pim/Bundle/InstallerBundle/Command/DumpRequirePathsCommand.php]: make protected

File: src/Pim/Bundle/NavigationBundle/Resources/public/js/navigation/abstract-view.js [hunk 15]
```diff
@@ -1,100 +1,104 @@
-define(['underscore', 'backbone', 'oro/navigation/dotmenu/view', 'pim/router'],
```
Comment (tamarasaurus) [inline src/Pim/Bundle/NavigationBundle/Resources/public/js/navigation/abstract-view.js]: delete

File: webpack.config.js [hunk 16]
```diff
@@ -0,0 +1,111 @@
+/* eslint-env es6 */
+
+const webpack = require('webpack')
+const path = require('path')
+const _ = require('lodash')
+
+const ContextReplacementPlugin = require('webpack/lib/ContextReplacementPlugin')
+const requirePaths = require(path.resolve('web/js/require-paths'))
+const AddToContextPlugin = require('./frontend/add-context-plugin')
+const pathOverrides = require('./frontend/path-overrides')
```
Comment (tamarasaurus) [inline webpack.config.js]: put in yaml

File: webpack.config.js [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+    entry: path.resolve(__dirname, './src/Pim/Bundle/EnrichBundle/Resources/public/js/index.js'),
+    output: {
+        path: path.resolve('./web/dist/'),
+        publicPath: '/dist/',
+        filename: 'app.min.js',
+        chunkFilename: '[name].bundle.js',
+        pathinfo: true,
+        devtoolLineToLine: true
+    },
+    resolve: {
```
Comment (tamarasaurus) [inline webpack.config.js]: add source map

**Comment by jjanvier** [comment]
(friendly reminder: don't forget to add the ticket number to your commits)

and good luck for the PR :+1:

**Comment by tamarasaurus** [comment]
@jjanvier Thanks! :)

**Pull request merged**